xbmc/addons/Scraper.cpp

```diff
@@ -187,6 +190,67 @@ std::string CScraper::GetPathSettings()
   return stream.str();
 }
 
+std::string CScraper::GetPathSettingsAsJSON()
+{
+  if (!LoadSettings(false))
+    return "{}";
+
+  CXBMCTinyXML doc;
+  SettingsToXML(doc);
+  TiXmlElement* root = doc.RootElement();
+
+  rapidjson::Document json_settings;
+  rapidjson::Document::AllocatorType& allocator = json_settings.GetAllocator();
+  json_settings.SetObject();
+  if (root)
+  {
+    for (TiXmlNode* node = root->FirstChild(); node; node = node->NextSibling())
+    {
+      TiXmlElement* element = node->ToElement();
+      const char* id = element->Attribute("id");
+      if (id)
+      {
+        const char* str_value = element->GetText();
+        rapidjson::Document json_value;
+        json_value.Parse(str_value);
```

Comment:
Parsing a string and then checking the type with `Is...` is wrong, because `true`, `false`, `1` or `null` could be string.
Also enums are still mapped to a string and not to an object or array.

Correct is to map by the setting type.

Reply:
Setting types are not stored along with settings. So it's either type checking or passing everything as strings.

Reply:
Yes the type isn't stored in `settings.xml`, but that doesn't mean the settings have no type. Maybe converting the xml settings, which miss information, into json is the wrong approach.

Reply:
I don't see many volunteers to fix this issue, and this is as much as I can do. I actually asked @notspiff and he is OK with this approach. The only thing we need to decide if we need this type conversion. I'd rather pass everything as strings and let scraper developers to decide how to process those settings. @ronie what do you think?

Reply:
I'm fine with passing everything as string, but converting it to a type which might not match the actual setting type isn't acceptable.

Reply:
> I don't see many volunteers to fix this issue, and this is as much as I can do.

yup, i fully understand.
perhaps @Montellese is willing to provide some pointers on how to implement it properly.
if not, i can accept we'll have to live it it.

Reply:
Instead of relying on `SettingsToXml()` to do the work for you you could write a `SettingsToJson()` method which goes through the whole settings object tree and converts the settings properly to JSON. It shouldn't be too hard but I admit that it was a mistake to include the XML serialization / deserialization in the settings classes / code.

Reply:
I see 2 problems:
- Real settings processing is buried deeply in functions call chain, far from Python scraper calls. I don't like the idea of creating similar deep call chain.
- I did not find methods to covert settings to their native types, only `ToString()`, so it seems that I'd have to rely on `std::stoi` and similar functions and I don't like this approach.

Since I'm no expert in Kodi core codebase and C++ in general, I'll appreciate any suggestions, but if I'm not wrong, I'd rather stick to the current approach.

Reply:
The "proper" approach would be to extract the XML serialization from the whole settings library and provide it as a serialization implementation (based on a generic interface). Then the same interface could be implemented for JSON.
Yes `CSetting` only has a `ToString()` method but using `CSetting::GetType()` you can figure out to which derived type you can / must cast the `CSetting` object and then use `GetValue()` to get the value with the proper type.

Reply:
I think this should go in as is, and should be backported to 18.6.

Passing settings as an arg is not ideal anyway, it was something quick that could solve the biggest problem while still playing nice with all of the existing scraper behavior. "Proper" isn't really useful in this context, particularly if it will prevent fixing this for Leia, the first version Python scrapers were introduced.